### PR TITLE
Refactor method parsing for clearer and faster code

### PR DESCRIPTION
**Pull request by dolmen**
I've refactored the code that parses HTTP methods to replace with a single switch the multiple if/else in multiple depth. The code is now much easier to read and faster.
In the second commit, the parsing for methods PUT/PATCH is put at the top of the table as those requests are much more common than the other methods in this second level dispatch.

### CODE REVIEW

File: http_parser.c [hunk 1]
```diff
@@ -1000,89 +1000,39 @@ size_t http_parser_execute (http_parser *parser,
...
-            }
-          } else {
-            SET_ERRNO(HPE_INVALID_METHOD);
-            goto error;
           }
-        } else if (parser->index == 4 && parser->method == HTTP_PROPFIND && ch == 'P') {
-          parser->method = HTTP_PROPPATCH;
-        } else if (parser->index == 3 && parser->method == HTTP_UNLOCK && ch == 'I') {
-          parser->method = HTTP_UNLINK;
+        } else if (ch == '-' && parser->index == 1
```
Comment (indutny) [inline http_parser.c]: Please put the operator before the newline.

**Comment by indutny** [comment]
Wow, this is actually looking very interesting! I will try to review it in a day or two. Thanks!

**Comment by dolmen** [comment]
ping!

**Comment by motionx77** [comment]
The code looks good to me. I'll test the speed tomorrow, on my server.

**Comment by paulofaria** [comment]
:+1:

**Comment by jasnell** [comment]
Oh I missed this one! Excellent, will give it a review as soon as possible.

**Comment by dolmen** [comment]
ping!

**Comment by jasnell** [comment]
haven't forgotten! just haven't had a spare moment :-/ ...@indutny ping!

**Comment by jbergstroem** [comment]
This looks sweet! tests passes/compiles on a few platforms I tried on.

File: http_parser.c [hunk 2]
```diff
@@ -1000,89 +1000,40 @@ size_t http_parser_execute (http_parser *parser,
...
+            case (HTTP_##meth << 16 | pos << 8 | ch): \
+              parser->method = HTTP_##new_meth; break;
+
+            XX(POST,      1, 'U', PUT)
+            XX(POST,      1, 'A', PATCH)
+            XX(CONNECT,   1, 'H', CHECKOUT)
+            XX(CONNECT,   2, 'P', COPY)
+            XX(MKCOL,     1, 'O', MOVE)
+            XX(MKCOL,     1, 'E', MERGE)
+            XX(MKCOL,     2, 'A', MKACTIVITY)
```
Comment (indutny) [inline http_parser.c]: Why can't `MSEARCH` be here too?

File: http_parser.c [hunk 3]
```diff
@@ -1000,89 +1000,40 @@ size_t http_parser_execute (http_parser *parser,
...
-              SET_ERRNO(HPE_INVALID_METHOD);
-              goto error;
-            }
-          } else if (parser->method == HTTP_UNLOCK) {
-            if (ch == 'S') {
-              parser->method = HTTP_UNSUBSCRIBE;
-            } else if(ch == 'B') {
-              parser->method = HTTP_UNBIND;
-            } else {
+        } else if (IS_ALPHA(ch)) {
```
Comment (indutny) [inline http_parser.c]: Ah, because of this, I guess?

File: http_parser.c [hunk 4]
```diff
@@ -1000,89 +1000,40 @@ size_t http_parser_execute (http_parser *parser,
...
+
+          switch (parser->method << 16 | parser->index << 8 | ch) {
```
Comment (indutny) [inline http_parser.c]: And this... gotcha

**Comment by indutny** [comment]
LGTM, but going to run benchmarks first. Thank you very much for this!

**Comment by indutny** [comment]
Benchmarks are happy, and even a bit faster (thought, it could be just a noise and not actual improvement).

**Comment by indutny** [comment]
Landed in fd65b0f, thank you!

**Comment by indutny** [comment]
And I'm deeply sorry for delaying it so much. This is very unfortunate. Thank you for the contribution!

File: http_parser.c (hunk 2 above)
Comment (dolmen) [inline http_parser.c]: Because this is `M-SEARCH`. We have to handle the dash differently.

**Comment by dolmen** [comment]
:clap:

**Pull request closed without merging**